subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/IncrementalInputChanges.java

```diff
@@ -40,18 +40,21 @@ public boolean isIncremental() {
 
     @Override
     public Iterable<InputFileDetails> getFileChanges(Object property) {
-        String propertyName = determinePropertyName(property, propertyNameByValue);
+        String propertyName = determinePropertyName(property, propertyNamesByValue);
         CollectingChangeVisitor visitor = new CollectingChangeVisitor();
         changes.accept(propertyName, visitor);
         return Cast.uncheckedNonnullCast(visitor.getChanges());
     }
 
-    public static String determinePropertyName(Object property, Map<Object, String> propertyNameByValue) {
-        String propertyName = propertyNameByValue.get(property);
-        if (propertyName == null) {
-            throw new UnsupportedOperationException("Cannot query incremental changes: No property found for " + property + ".");
+    public static String determinePropertyName(Object propertyValue, ImmutableListMultimap<Object, String> propertyNameByValue) {
+        ImmutableList<String> propertyNames = propertyNameByValue.get(propertyValue);
+        if (propertyNames.isEmpty()) {
+            throw new UnsupportedOperationException("Cannot query incremental changes: No property found for value " + propertyValue + ".");
+        }
+        if (propertyNames.size() > 1) {
+            throw new UnsupportedOperationException(String.format("Cannot query incremental changes: More that one property found with value %s: %s.", propertyValue, propertyNames));
```

Comment:
Can't we return all the changes belonging to each property in this case? What's a use case where this happens?

Reply:
It happens when a task does not use the `Property` API, yet, and has the same value in both parameters. If the task is using the `Property` API, then each parameter will be a separate instance and the given case cannot happen.

We could silently return one of the changes or both, it is still probably unexpected for the user, so I prefer to fail if we cannot determine the input property uniquely. It gets especially complicated when the two parameters have different normalization.

Reply:
FTR, this won't be necessary when we mark incremental input properties with an annotation.